yapapi/executor/services.py

```diff
@@ -294,53 +294,72 @@ def _get_handler(instance: ServiceInstance):
-            # * we may be losing a `batch` when we act on the control signal
-            #
-            # potential solution:
-            # * use `aiostream.stream.merge`
-
-            ctl = instance.get_control_signal()
-            if ctl == ControlSignal.stop:
-                if instance.state == ServiceState.running:
-                    instance.service_state.stop()
-                else:
-                    instance.service_state.terminate()
-
-                logger.debug(f"{instance.service} state changed to {instance.state.value}")
-
-                handler = self._get_handler(instance)
-                batch = None
+            # Repeatedly wait on one of `(batch_task, signal_task)` to finish.
+            # If it's the first one, retrieve a batch from its result and handle it.
+            # If it's the second -- retrieve and handle a signal.
+            # Any finished task is replaced with a new one, so there are always two.
+
+            if batch_task is None:
+                batch_task = loop.create_task(handler.__anext__())
+            if signal_task is None:
+                signal_task = loop.create_task(instance.control_queue.get())
+
+            done, _ = await asyncio.wait(
+                (batch_task, signal_task), return_when=asyncio.FIRST_COMPLETED
+            )
+
+            if batch_task in done:
+                # Process a batch
+                try:
+                    batch = batch_task.result()
+                    fut_result = yield batch
+                    result = await fut_result
+                    wrapped_results = loop.create_future()
+                    wrapped_results.set_result(result)
+                    batch_task = loop.create_task(handler.asend(wrapped_results))
+                except StopAsyncIteration:
+                    instance.service_state.lifecycle()
+                    handler = self._get_handler(instance)
+                    logger.debug(f"{instance.service} state changed to {instance.state.value}")
+                    batch_task = None
+
+            if signal_task in done:
+                # Process a signal
+                ctl = signal_task.result()
+                logger.debug(f"Processing control signal {ctl}")
+                if ctl == ControlSignal.stop:
+                    if instance.state == ServiceState.running:
+                        instance.service_state.stop()
+                    else:
+                        instance.service_state.terminate()
+                    handler = self._get_handler(instance)
+                    if batch_task:
+                        batch_task.cancel()
+                        batch_task = None
+                signal_task = None
 
         logger.info(f"{instance.service} decomissioned")
 
+        if batch_task:
+            batch_task.cancel()
+            await batch_task
+        if signal_task:
+            signal_task.cancel()
+            await signal_task
+
```

Comment:
I think this should go _before_ the above logger message, and maybe a new `.debug` one should be added above

Reply:
done, https://github.com/golemfactory/yapapi/pull/408/commits/648c114eb28c5db3ad445b8cb837105645d75a8a